homeassistant/components/sensor/__init__.py

```diff
@@ -127,3 +134,75 @@ def state_attributes(self) -> dict[str, Any] | None:
             return {ATTR_LAST_RESET: last_reset.isoformat()}
 
         return None
+
+    @property
+    def native_value(self) -> StateType:
+        """Return the value reported by the sensor."""
+        return self._attr_native_value
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the sensor, if any."""
+        return self._attr_native_unit_of_measurement
+
+    @property
+    def unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement of the entity, after unit conversion."""
+        if self._attr_unit_of_measurement is not None:
+            return self._attr_unit_of_measurement
+
+        native_unit_of_measurement = self.native_unit_of_measurement
+
+        if native_unit_of_measurement in (TEMP_CELSIUS, TEMP_FAHRENHEIT):
+            # Suppress AttributeError to handle tests testing directly on entity objects
+            with suppress(ValueError, AttributeError):
+                return self.hass.config.units.temperature_unit
+
+        return native_unit_of_measurement
+
+    @property
+    def state(self) -> Any:
+        """Return the state of the sensor and perform unit conversions, if needed."""
+        if self._attr_state is not None:
+            return self._attr_state
+
+        unit_of_measurement = self.native_unit_of_measurement
+        value = self.native_value
+
+        # Handle tests testing directly on entity objects without setting hass
+        if not self.hass:
+            return value
```

Comment:
This looks like we're introducing incorrect behavior to handle incorrect tests. We always want to convert the state if the device class / unit matches the conversion criteria.

If this is really only a condition that happens in tests, I'd say it's better to remove or disable those tests.

Reply:
Yes, you're right, we should not fix it here.
I'll try to fix the offending tests, and if not trivial disable them.

Reply:
Fixed in 25767ec